// @balticcode/ngx-hotkeys/esm5/lib/interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function IHotkey() { }
function IHotkey_tsickle_Closure_declarations() {
    /**
     * Key binding
     * @type {?}
     */
    IHotkey.prototype.combo;
    /**
     * Handler to call when binding is triggered
     * @type {?}
     */
    IHotkey.prototype.handler;
    /**
     * The type of event to listen for (for mousetrap)
     * @type {?|undefined}
     */
    IHotkey.prototype.specificEvent;
    /**
     * An array of tag names to allow this combo in ('INPUT', 'SELECT', and/or 'TEXTAREA')
     * @type {?|undefined}
     */
    IHotkey.prototype.allowIn;
    /**
     * Description for the help menu
     * @type {?|undefined}
     */
    IHotkey.prototype.description;
    /**
     * Custom display format used in cheatsheet
     * @type {?|undefined}
     */
    IHotkey.prototype.format;
}
/**
 * @record
 */
export function IHotkeyOptions() { }
function IHotkeyOptions_tsickle_Closure_declarations() {
    /**
     * Disable the cheat sheet popover dialog? Default: false
     * @type {?|undefined}
     */
    IHotkeyOptions.prototype.disableCheatSheet;
    /**
     * Specifies the title of the cheat sheet. Default: 'Keyboard Shortcuts:'
     * @type {?|undefined}
     */
    IHotkeyOptions.prototype.cheatSheetTitle;
    /**
     * Key combination to trigger the cheat sheet. Default: '?'
     * @type {?|undefined}
     */
    IHotkeyOptions.prototype.cheatSheetHotkey;
    /**
     * Description for the cheat sheet hot key in the cheat sheet. Default: 'Show / hide this help menu'
     * @type {?|undefined}
     */
    IHotkeyOptions.prototype.cheatSheetHotkeyDescription;
    /**
     * Use also ESC for closing the cheat sheet. Default: false
     * @type {?|undefined}
     */
    IHotkeyOptions.prototype.cheatSheetCloseEsc;
    /**
     * Description for the ESC key for closing the cheat sheet (if enabled). Default: 'Hide this help menu'
     * @type {?|undefined}
     */
    IHotkeyOptions.prototype.cheatSheetCloseEscDescription;
}
/**
 * @record
 */
export function ExtendedKeyboardEvent() { }
function ExtendedKeyboardEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    ExtendedKeyboardEvent.prototype.returnValue;
}
/**
 * @record
 */
export function HotKeyMap() { }
function HotKeyMap_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [combo: string]: (event: KeyboardEvent, combo: string) => ExtendedKeyboardEvent | boolean;
    */
}
