// @balticcode/ngx-hotkeys/esm2015/lib/ngx-hotkeys.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxCheatsheetComponent } from './ngx-cheatsheet/ngx-cheatsheet.component';
import { NgxHotkeysDirective } from './ngx-hotkeys.directive';
import { NgxHotkeysService } from './ngx-hotkeys.service';
import { HotkeyOptions } from './token';
export class NgxHotkeysModule {
    /**
     * @param {?=} options
     * @return {?}
     */
    static forRoot(options) {
        return {
            ngModule: NgxHotkeysModule,
            providers: [
                NgxHotkeysService,
                { provide: HotkeyOptions, useValue: options }
            ]
        };
    }
}
NgxHotkeysModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [NgxCheatsheetComponent, NgxHotkeysDirective],
                exports: [NgxHotkeysDirective, NgxCheatsheetComponent],
                providers: [NgxHotkeysService]
            },] },
];
function NgxHotkeysModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxHotkeysModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxHotkeysModule.ctorParameters;
}
