// @balticcode/ngx-hotkeys/esm2015/lib/ngx-cheatsheet/ngx-cheatsheet.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { NgxHotkeysService } from '../ngx-hotkeys.service';
export class NgxCheatsheetComponent {
    /**
     * @param {?} hotkeysService
     */
    constructor(hotkeysService) {
        this.hotkeysService = hotkeysService;
        this.helpVisible = false;
        this._map = {
            command: '\u2318',
            // ⌘
            shift: '\u21E7',
            // ⇧
            left: '\u2190',
            // ←
            right: '\u2192',
            // →
            up: '\u2191',
            // ↑
            down: '\u2193',
            // ↓
            'return': '\u23CE',
            // ⏎
            backspace: '\u232B' // ⌫
        };
        this.title = this.hotkeysService.options.cheatSheetTitle;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._subscription = this.hotkeysService.cheatSheetToggled
            .subscribe((isOpen) => {
            if (isOpen !== false) {
                this.hotkeys = this.hotkeysService.hotkeys
                    .filter(hotkey => hotkey.description);
            }
            if (isOpen === false) {
                this.helpVisible = false;
            }
            else {
                this.toggleCheatSheet();
            }
        });
    }
    /**
     * @param {?} hotkey
     * @return {?}
     */
    formatHotkey(hotkey) {
        if (!hotkey.format) {
            const /** @type {?} */ combo = Array.isArray(hotkey.combo) ? hotkey.combo[0] : hotkey.combo;
            const /** @type {?} */ sequence = combo.split(/[\s]/);
            for (let /** @type {?} */ i = 0; i < sequence.length; i++) {
                sequence[i] = this.symbolize(sequence[i]);
            }
            hotkey.format = sequence;
        }
        return hotkey.format;
    }
    /**
     * @param {?} combo
     * @return {?}
     */
    symbolize(combo) {
        const /** @type {?} */ comboSplit = combo.split('+');
        for (let /** @type {?} */ i = 0; i < comboSplit.length; i++) {
            // try to resolve command / ctrl based on OS:
            if (comboSplit[i] === 'mod') {
                if (window.navigator && window.navigator.platform.indexOf('Mac') >= 0) {
                    comboSplit[i] = 'command';
                }
                else {
                    comboSplit[i] = 'ctrl';
                }
            }
            comboSplit[i] = this._map[comboSplit[i]] || comboSplit[i];
        }
        return comboSplit.join(' + ');
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._subscription) {
            this._subscription.unsubscribe();
        }
    }
    /**
     * @return {?}
     */
    toggleCheatSheet() {
        this.helpVisible = !this.helpVisible;
    }
}
NgxCheatsheetComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-cheatsheet',
                template: `<div class="ngx-cheatsheet fade" [ngClass]="{'in': helpVisible}" style="display:none">
  <div class="ngx-cheatsheet__hotkeys">
    <h4 class="ngx-cheatsheet__hotkeys__title">{{ title }}</h4>
    <table>
      <tbody>
      <tr *ngFor="let hotkey of hotkeys">
        <td class="ngx-cheatsheet__hotkeys__keys">
          <span class="ngx-cheatsheet__hotkeys__keys__key">{{ formatHotkey(hotkey) }}</span>
        </td>
        <td class="ngx-cheatsheet__hotkeys__keys__text">{{ hotkey.description }}</td>
      </tr>
      </tbody>
    </table>
    <div class="ngx-cheatsheet__hotkeys__close" (click)="toggleCheatSheet()">&#215;</div>
  </div>
</div>
`,
                styles: [`.ngx-cheatsheet{display:table!important;position:fixed;width:100%;height:100%;top:0;left:0;color:#333;font-size:1em;background-color:rgba(255,255,255,.9)}.ngx-cheatsheet.fade{z-index:-1024;visibility:hidden;opacity:0;transition:opacity .15s linear}.ngx-cheatsheet.fade.in{z-index:10002;visibility:visible;opacity:1}.ngx-cheatsheet__hotkeys{width:100%;height:100%;display:table-cell;vertical-align:middle}.ngx-cheatsheet__hotkeys__title{font-weight:700;text-align:center;font-size:1.2em}.ngx-cheatsheet__hotkeys table{margin:auto;color:#333}.ngx-cheatsheet__hotkeys__keys{padding:5px;text-align:right}.ngx-cheatsheet__hotkeys__keys__key{display:inline-block;color:#fff;background-color:#333;border:1px solid #333;border-radius:5px;text-align:center;margin-right:5px;box-shadow:inset 0 1px 0 #666,0 1px 0 #bbb;padding:5px 9px;font-size:1em}.ngx-cheatsheet__hotkeys__keys__text{padding-left:10px;font-size:1em}.ngx-cheatsheet__hotkeys__close{position:fixed;top:20px;right:20px;font-size:2em;font-weight:700;padding:5px 10px;border:1px solid #ddd;border-radius:5px;min-height:45px;min-width:45px;text-align:center}@media all and (max-width:500px){.ngx-cheatsheet__hotkeys{font-size:.8em}}@media all and (min-width:750px){.ngx-cheatsheet__hotkeys{font-size:1.2em}}`]
            },] },
];
/** @nocollapse */
NgxCheatsheetComponent.ctorParameters = () => [
    { type: NgxHotkeysService, },
];
NgxCheatsheetComponent.propDecorators = {
    "title": [{ type: Input },],
};
function NgxCheatsheetComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxCheatsheetComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxCheatsheetComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NgxCheatsheetComponent.propDecorators;
    /** @type {?} */
    NgxCheatsheetComponent.prototype.title;
    /** @type {?} */
    NgxCheatsheetComponent.prototype.helpVisible;
    /** @type {?} */
    NgxCheatsheetComponent.prototype.hotkeys;
    /** @type {?} */
    NgxCheatsheetComponent.prototype._subscription;
    /** @type {?} */
    NgxCheatsheetComponent.prototype._map;
    /** @type {?} */
    NgxCheatsheetComponent.prototype.hotkeysService;
}
